src/actions/AccessControl.ts

```diff
@@ -25,42 +50,30 @@ export const createSessionKeyTransactionParams = async (
             "https://docs.fun.xyz"
         )
     }
+    if (params.userId === undefined) {
+        throw new InvalidParameterError(
+            ErrorCode.MissingParameter,
+            "userId is required",
+            { params },
+            "Provide userId when creating a session key.",
+            "https://docs.fun.xyz"
+        )
+    }
     let { actionValueLimit, feeValueLimit } = params
     actionValueLimit ??= 0n
     feeValueLimit ??= 0n
-    const selectors: Hex[] = []
-    params.actionWhitelist.forEach((actionWhitelistItem) => {
-        if (typeof actionWhitelistItem === "string") {
-            selectors.push(actionWhitelistItem)
-        } else {
-            selectors.push(
-                ...actionWhitelistItem.functionWhitelist.map((functionName) => getSigHash(actionWhitelistItem.abi, functionName))
-            )
-        }
-    })
-    const targets = params.targetWhitelist.map((target) => getAddress(target))
-
-    let feeRecipientTokenMerkleRootHash = HashOne
-    if (params.feeRecipientWhitelist && params.feeTokenWhitelist) {
-        const recipients = params.feeRecipientWhitelist.map((recipient) => getAddress(recipient))
-        const tokens = await Promise.all(params.feeTokenWhitelist.map((token) => Token.getAddress(token)))
-        const feeRecipientAndTokenMerkleTree = new MerkleTree([...recipients, ...tokens])
-        feeRecipientTokenMerkleRootHash = feeRecipientAndTokenMerkleTree.getRoot()
-        params.user.setFeeRecipientMerkleTree(feeRecipientAndTokenMerkleTree)
-    }
-    const targetSelectorMerkleTree = new MerkleTree([...targets, ...selectors])
-    const targetSelectorMerkleRootHash = targetSelectorMerkleTree.getRoot()
-    params.user.setTargetSelectorMerkleTree(targetSelectorMerkleTree)
+    const targetSelectorMerkleTree = await createTargetSelectorMerkleTree(params)
+    const feeRecipientTokenMerkleTree = await createFeeRecipientAndTokenMerkleTree(params)
     const ruleStruct: RuleStruct = {
         deadline: convertTimestampToBigInt(params.deadline),
-        targetSelectorMerkleRootHash,
-        feeRecipientTokenMerkleRootHash,
+        targetSelectorMerkleRootHash: targetSelectorMerkleTree.getRootHash(),
+        feeRecipientTokenMerkleRootHash: feeRecipientTokenMerkleTree.getRootHash(),
         actionValueLimit,
         feeValueLimit
     }
-    const roleId = params.user.roleId
-    const ruleId = params.user.ruleId
-    const userid = await params.user.getUserId()
+    const roleId = params.roleId
+    const ruleId = params.ruleId
+    const userid = params.userId
```

Comment:
Style suggestion

```suggestion
    const { roleId, ruleId, userId } = params
```

Related: I assume it's a typo that `userid` is not camelcase

Reply:
fixed!